File: src/main/java/com/handwoong/everyonewaiter/exception/ApiExceptionHandler.java
package com.handwoong.everyonewaiter.exception;

import lombok.extern.slf4j.Slf4j;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

@Slf4j
@RestControllerAdvice("com.handwoong.everyonewaiter.api")
public class ApiExceptionHandler {

    @ExceptionHandler({CustomException.class})
    public Object handleCustomException(CustomException err) {
        ErrorCode errorCode = err.getErrorCode();
        log.error("[{}] {} {}", errorCode.name(), errorCode.getStatus(), errorCode.getMessage());
        return ErrorResponse.toResponseEntity(errorCode);
    }

    @ExceptionHandler({MethodArgumentNotValidException.class})
    public Object handleCustomException(MethodArgumentNotValidException err) {
        ErrorCode errorCode = ErrorCode.METHOD_ARGUMENT_NOT_VALID;
        log.error("[{}] {} {} {}", errorCode.name(), errorCode.getStatus(), errorCode.getMessage(),
                err.getDetailMessageArguments());
        return ErrorResponse.toResponseEntity(errorCode);
    }

    @ExceptionHandler({Exception.class})
    public Object handleException(Exception err) {
        ErrorCode errorCode = ErrorCode.INTERNAL_SERVER_ERROR;
        log.error("[{}] {} {} {}",
                errorCode.name(), errorCode.getStatus(), errorCode.getMessage(), err.getMessage());
        return ErrorResponse.toResponseEntity(errorCode);
    }
}

File: src/main/kotlin/com/handwoong/everyonewaiter/exception/ApiExceptionHandler.kt
package com.handwoong.everyonewaiter.exception

import com.handwoong.everyonewaiter.util.logger
import org.springframework.http.ResponseEntity
import org.springframework.security.authentication.BadCredentialsException
import org.springframework.security.authentication.InternalAuthenticationServiceException
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice

@RestControllerAdvice
class ApiExceptionHandler {

    private val log = logger()

    @ExceptionHandler(value = [CustomException::class])
    fun handleCustomException(error: CustomException): ResponseEntity<ErrorResponse> {
        val errorCode = error.errorCode
        log.error("[${errorCode.name}] ${errorCode.status} ${errorCode.message}")
        return ErrorResponse.of(errorCode)
    }

    @ExceptionHandler(value = [MethodArgumentNotValidException::class])
    fun handleNotValidException(error: MethodArgumentNotValidException): ResponseEntity<ErrorResponse> {
        val errorCode = ErrorCode.METHOD_ARGUMENT_NOT_VALID
        log.error("[${errorCode.name}] ${errorCode.status} ${errorCode.message} ${error.message}")
        return ErrorResponse.of(errorCode)
    }

    @ExceptionHandler(value = [BadCredentialsException::class])
    fun handleBadCredentialsException(error: BadCredentialsException): ResponseEntity<ErrorResponse> {
        val errorCode = ErrorCode.CHECK_USERNAME_AND_PASSWORD
        log.error("[${errorCode.name}] ${errorCode.status} ${errorCode.message}")
        return ErrorResponse.of(errorCode)
    }

    @ExceptionHandler(value = [InternalAuthenticationServiceException::class])
    fun handleInternalAuthenticationServiceException(error: InternalAuthenticationServiceException): ResponseEntity<ErrorResponse> {
        val errorCode = ErrorCode.CHECK_USERNAME_AND_PASSWORD
        log.error("[${errorCode.name}] ${errorCode.status} ${errorCode.message}")
        return ErrorResponse.of(errorCode)
    }

    @ExceptionHandler(value = [Exception::class])
    fun handleException(error: Exception): ResponseEntity<ErrorResponse> {
        val errorCode = ErrorCode.SERVER_ERROR
        log.error("[${errorCode.name}] ${errorCode.status} ${errorCode.message}")
        log.error("${error}")
        return ErrorResponse.of(errorCode)
    }

}
